boolbv_width: distinguish zero from unknown size

tautschnig opened the pull request:
Use optionalt to distinguish types of known-zero size from those with
unknown size, which previously had the default value of zero.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening commented:
I got somewhat confused by the fact that a boolbv_widtht cache entry no longer has type `entryt` but `optional<entryt>`.

tautschnig commented:
> I got somewhat confused by the fact that a boolbv_widtht cache entry no longer has type `entryt` but `optional<entryt>`.

If it weren't `optionalt`, then we'd either need a Boolean flag in there ("invalid") or else types that don't have a width couldn't be cached. I'm happy to adjust the implementation to do either of these if that's deemed easier to follow.

kroening commented:
> If it weren't `optionalt`, then we'd either need a Boolean flag in there ("invalid") or else types that don't have a width couldn't be cached. I'm happy to adjust the implementation to do either of these if that's deemed easier to follow.

How about s.th. like `using entryt = optionalt<entry_structt>`

tautschnig commented:
> > If it weren't `optionalt`, then we'd either need a Boolean flag in there ("invalid") or else types that don't have a width couldn't be cached. I'm happy to adjust the implementation to do either of these if that's deemed easier to follow.
> 
> How about s.th. like `using entryt = optionalt<entry_structt>`

Done (I chose to use `defined_entryt` instead of `entry_structt`).

peterschrammel commented on src/solvers/flattening/boolbv.cpp:
```diff
@@ -85,8 +85,11 @@ bvt boolbvt::conversion_failed(const exprt &expr)
   // try to make it free bits
-  std::size_t width=boolbv_width(expr.type());
-  return prop.new_variables(width);
+  // TODO we likely end up in this path when an earlier call to boolbv_width
```
Does the TODO mean that something needs to be done or that we don't understand why this piece of code exists?

tautschnig replied:
Thank you for reminding me of this 😊 I have now audited all uses of `conversion_failed`, several of which were in response to calls to `boolbv_width` failing. That, however, would have resulted in the call to `boolbv_width` in here failing, making the code path really dangerous in the way it previously was (we would not have generated the correct number of free variables), and useless the way I changed it (making the `CHECK_RETURN` in here fail). All of these instances are now turned into early failures.

tautschnig committed:
boolbv_width: distinguish zero from unknown size

Use optionalt to distinguish types of known-zero size from those with
unknown size, which previously had the default value of zero.

This makes it possible to support index expressions over zero bitwidth
arrays without mistaking the situation for an unknown size. (There is
not really anything wrong in having empty bitvectors, which we otherwise
already support (as of e021eef4).)
